### PR TITLE
feat(binance): add support for websocket trading

### PR DESCRIPTION
**Pull request by pcriadoperez**
- Add support for WebSocket API and `createOrderWs`, `editOrderWs`, `cancelOrderWs`, `cancelAllOrdersWs`, `fetchOrderWs`, `fetchOrdersWs`, `fetchBalanceWs`, `fetchTradesWs`
- Move `createOrder` request logic into `createOrderRequest` to be used also by websocket API 
- Move `editSpotOrder` request logic into `editSpotOrderRequest` to be used also by websocket API 
- Update Spot errors

Note: No logic was added to the rateLimiter, and the default one is used. However in the future it could be optmized to also add weights to each ws request.

### CODE REVIEW

File: ts/src/binance.ts [hunk 1]
```diff
@@ -3693,108 +3697,6 @@ export default class binance extends Exchange {
         const clientOrderId = this.safeStringN (params, [ 'newClientOrderId', 'clientOrderId', 'origClientOrderId' ]);
         let response = undefined;
```
Comment (carlosmiei) [inline ts/src/binance.ts:3698]: @pcriadoperez should we move this logic inside `createOrderRequest` ?
image

File: ts/src/pro/binance.ts [hunk 2]
```diff
@@ -1280,6 +1395,446 @@ export default class binance extends binanceRest {
...
+         * @method
+         * @name binance#fetchOrderWs
+         * @see https://binance-docs.github.io/apidocs/websocket_api/en/#query-order-user_data
+         * @description fetches information on an order made by the user
+         * @param {string} symbol unified symbol of the market the order was made in
+         * @param {object} params extra parameters specific to the binance api endpoint
+         * @returns {object} An [order structure]{@link https://docs.ccxt.com/#/?id=order-structure}
+         */
+        await this.loadMarkets ();
+        this.checkIsSpot ('fetchOrderWs', symbol);
```
Comment (carlosmiei) [inline ts/src/pro/binance.ts:1739]: @pcriadoperez `checkIsSpot` assumes that `symbol` is defined but that might not be the case here

Comment (pcriadoperez) [reply ts/src/pro/binance.ts:1739]: Good catch, in this case the api actually requires the symbol, but updated the function signature and checked the others. Also updated checkIsSpot.

File: ts/src/pro/binance.ts [hunk 3]
```diff
@@ -1280,6 +1395,446 @@ export default class binance extends binanceRest {
...
+         * @returns {object} an list of [order structures]{@link https://docs.ccxt.com/en/latest/manual.html#order-structure}
+         */
+        await this.loadMarkets ();
+        this.checkIsSpot ('cancelOrderWs', symbol);
+        const url = this.urls['api']['ws']['ws'];
+        const requestId = this.requestId (url);
+        const messageHash = requestId.toString ();
+        let returnRateLimits = false;
+        [ returnRateLimits, params ] = this.handleOptionAndParams (params, 'cancelOrderWs', 'returnRateLimits', false);
+        const payload = {
```
[the remark this answers is not in the document]
Comment (pcriadoperez) [reply ts/src/pro/binance.ts:1683]: Good catch, I extend the params

File: ts/src/pro/binance.ts (hunk 2 above)
Comment (pcriadoperez) [reply ts/src/pro/binance.ts:1750]: I made it undefined and add a BadRequest for the functions where symbol is required

**Review by carlosmiei — APPROVED** [review]

**Pull request merged**